### PR TITLE
backport optional extension building for python 2.7

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this sorts out the issues i caused on python2.7 by backporting the approach of the python3 optional parameter to python2

this fixes regressions introduced in #106

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -2,12 +2,15 @@
 
 try:
     from setuptools import setup, Extension
+    from setuptolls.command.build_ext import build_ext as base_build_ext
```
Comment (lorengordon) [inline setup.py]: This should be `setuptools`, no? (Not `tolls`...)

Comment (RonnyPfannschmidt) [reply setup.py]: yikes, yes, im going to put in a warning with the exception when it fails

File: setup.py [hunk 2]
```diff
@@ -2,12 +2,20 @@
...
 try:
     from setuptools import setup, Extension
+    from setuptools.command.build_ext import build_ext as base_build_ext
 except ImportError:
-    from distutils.core import setup, Extension
+    import warnings
+    import sys
+    val = sys.exc_info()[1]
 
+    warnings.warn("import of setuptools failed %r" % val)
```
Comment (benhoyt) [inline setup.py]: I don't think this warning is necessary -- it falls back to distutils anyway. It seems to me this was added to debug a spelling mistake in an earlier commit. Now that we're importing it correctly, let's just leave this out.

File: setup.py [hunk 3]
```diff
@@ -25,8 +33,18 @@
     long_description = f.read()
 
 
-# the extension is optional since in case of lack of c the api
-# there is a ctypes fallback and a slow python fallback
+class BuildExt(base_build_ext):
```
Comment (benhoyt) [inline setup.py]: Is this code copied from somewhere, the Python 3.x optional=true code or something? Just curious.

File: setup.py [hunk 4]
```diff
@@ -25,8 +33,18 @@
...
+    # there is a ctypes fallback and a slow python fallback
+
+    def build_extension(self, ext):
+        try:
+            base_build_ext.build_extension(self, ext)
+        except Exception:
+            exception = sys.exc_info()[0]
+            logging.warn("building the %s failed with %s", ext.name, exception)
+
 extension = Extension('_scandir', ['_scandir.c'], optional=True)
```
Comment (benhoyt) [inline setup.py]: Won't this still fail as we still always include the `optional=True` flag here? Or is it just a warning and so we're ignoring it?

Comment (RonnyPfannschmidt) [reply setup.py]: its just a warning on python2, i'm ignoring it here

the `build_ext command is what's interpreting those

File: setup.py (hunk 2 above)
Comment (RonnyPfannschmidt) [reply setup.py]: i'd like to leave it in so user see that setuptools is missing
from a packaging standpoint i'd like to completely remove the support for plain distutils if this was my project

File: setup.py (hunk 3 above)
Comment (RonnyPfannschmidt) [reply setup.py]: its a small reimplementation of the python3 mechanism, which has a context managers and a few more bits to sort it out

**Comment by nicoddemus** [comment]
Thanks @benhoyt for merging this. 👍 

Would you mind pinging this issue when a new release comes out? Thanks!

**Comment by benhoyt** [comment]
Done! It's version 1.9.0.

**Comment by nicoddemus** [comment]
Awesome, thank you!

**Pull request merged**